### PR TITLE
Add RPC Conflict Management

### PR DESCRIPTION
**Pull request by Akihiro-Miyazaki**
## Introduction
This proposal is to add a function to SDL Core that manages concurrent RPC conflicts according to priority. This will allow OEMs to receive the expected request from SDL Core during RPC conflict by modifying the configuration file in which the priority is set.

### CODE REVIEW

**Comment by Akihiro-Miyazaki** [comment]
Hi. @theresalech -san
Could you please review my proposal?
Thank you!

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
+# Add RPC Conflict Management
+
+* Proposal: [SDL-NNNN](NNNN-Clarification-of-SDL-App-Icon-display-sequence-and-display-order.md)
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:3]:
```suggestion
* Proposal: [SDL-NNNN](NNNN-Add-RPC-Conflict-Management.md)
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+* Author: [Akihiro Miyazaki (Nexty)](https://github.com/Akihiro-Miyazaki)
+* Status: **Awaiting review**
+* Impacted Platforms: [ Core / SDL Documents ]
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:6]: Please remove `SDL Documents` as this is not a possible platform per the proposal template: https://github.com/smartdevicelink/sdl_evolution/blob/master/0000-template.md.  If there are any other platforms in the template that should be included here, please add.

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+## Introduction
+This proposal is to add a function to SDL Core that manages concurrent RPC conflicts according to priority. This will allow OEMs to receive the expected request from SDL Core during RPC conflict by modifying the configuration file in which the priority is set.
+
+
+## Motivation
+When multiple apps operate, ON-screen (ONS) messages and text-to-speech (TTS) RPCs are requested together. However, the current SDL Core does not have the function to manage these conflicts. Thus, all requests are notified to the middleware and managed by the OEMs themselves.For instance, ONS and TTS conflicts are managed by prioritizing the latter RPC. The figure below demonstrates the sequence process when `PerformAudioPassThru`occurs during  `ScrollableMessage`
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:13]:
```suggestion
When multiple apps operate, ON-screen (ONS) messages and text-to-speech (TTS) RPCs are requested together. However, the current SDL Core does not have the function to manage these conflicts. Thus, all requests are notified to the middleware and managed by the OEMs themselves. For instance, ONS and TTS conflicts are managed by prioritizing the latter RPC. The figure below demonstrates the sequence process when `PerformAudioPassThru` occurs during `ScrollableMessage`.
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+![Figure1_Conflict_between_ScrollableMessage_and_PerformAudioPassThru.png](../assets/proposals/NNNN-Add-RPC-Conflict-Management/Figure1_Conflict_between_ScrollableMessage_and_PerformAudioPassThru.png)
+
+The example above shows that the latter RPC `PerformAudioPassThru` is displayed and aborts `ScrollableMessage`. However, this kind of method has the following problems.
+1. Since there is no guideline for RPC conflicts, implementations vary for each OEMs, thus, it is hard to standardize.
+2. Management of RPC conflicts with OEMs may increase the difficulty of implementing middleware.
+
+To solve these problems, we propose to add a new RPC conflict management function to SDL Core.
+
+## Proposed solution
+To implement the RPC conflict management, we will add a new RPC conflict management module, InterruptManager, and configuration file, InterruptManagerConfig, to SDL Core.
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:25]:
```suggestion
To implement the RPC conflict management, we will add a new RPC conflict management module, `InterruptManager`, and configuration file, `InterruptManagerConfig`, to SDL Core.
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+RPC priority table and AppHMIType priority table are specified in InterruptManagerConfig. RPC priority table and AppHMIType priority table are tables that each set the priority of RPC and app type (`appHMIType`). By modifying InterruptManagerConfig, the OEM can receive the expected request from SDL Core during RPC conflict. On the other hand, InterruptManager reads InterruptManagerConfig during the SDL Core startup and builds the two tables mentioned above based on their settings. When an RPC conflict occurs, the `InterruptManager` first determines the RPC with a high priority according to the RPC priority table. However, if two competing RPCs have the same priority, then the RPC with the higher priority is determined according to the AppHMIType priority table.
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:27]:
```suggestion
RPC priority table and `AppHMIType` priority table are specified in `InterruptManagerConfig`. RPC priority table and `AppHMIType` priority table are tables that each set the priority of RPC and app type (`appHMIType`). By modifying `InterruptManagerConfig`, an OEM can receive the expected request from SDL Core during RPC conflict. On the other hand, `InterruptManager` reads `InterruptManagerConfig` during the SDL Core startup and builds the two tables mentioned above based on their settings. When an RPC conflict occurs, the `InterruptManager` first determines the RPC with a high priority according to the RPC priority table. However, if two competing RPCs have the same priority, then the RPC with the higher priority is determined according to the `AppHMIType` priority table.
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+#### RPC Conflict Management Configuration File
+The following tables of InterruptManagerConfig are explained.
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:30]:
```suggestion
The following tables of explain the `InterruptManagerConfig`.
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+- RPC priority table
+- AppHMIType priority table
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:33]:
```suggestion
- `AppHMIType` priority table
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+<b>RPC Priority Table</b><br>
+The RPC priority table describes the priority of each RPC. When multiple RPCs occur at the same time, the RPC with the higher priority is determined according to the RPC priority table. The target RPCs in this proposal are ONS and TTS. ONS RPCs include such as `Alert` and `PerformInteraction`. The RPC priority table can set the priority for each ONS RPC. On the other hand, there are two types of TTS RPCs, TTS with ONS and TTS only. TTS with ONS RPCs shall follow the processed priority of the ONS RPC, while TTS only RPCs can set the priority individually.
+
+Below is the default settings of RPC priority table.
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:38]:
```suggestion
Below are the default settings of RPC priority table.
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+OEMs can modify the RPC priority table and adjust the RPC priority to their specifications.
+
+The table below shows how the RPC will be determined by the priority set during RPC conflict based on the default settings above (table 1).
+
+<b>Table 2.</b> Priority result of Table 1
+
+![Table_2_Priority_result_of_Table1.png](../assets/proposals/NNNN-Add-RPC-Conflict-Management/Table_2_Priority_result_of_Table1.png)
+
+(* 1): For RPCs with the same priority, the AppHMIType priority table, which is described later, will be used to determine the priority.
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:60]:
```suggestion
(* 1): For RPCs with the same priority, the `AppHMIType` priority table, which is described later, will be used to determine the priority.
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+    "UI.Alert": 2,
+    "UI.PerformAudioPassThru": 2,
+    "UI.PerformInteraction": 3,
+    "UI.ScrollableMessage": 3,
+    "UI.Slider": 3,
+    "TTS.SPEAK": 3
+}
+```
+
+OEMs can modify any RPC priority. For example, if the priority of UI.PerformInteraction is modified to "1", its RPC priority will be the same as `BC.DialNumber`.
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:75]:
```suggestion
OEMs can modify any RPC priority. For example, if the priority of `UI.PerformInteraction` is modified to "1", its RPC priority will be the same as `BC.DialNumber`.
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+	"UI.PerformAudioPassThru": 2,
+	"UI.PerformInteraction": 3,
+	"UI.ScrollableMessage": 3,
+	"UI.Slider": 2,
+	"TTS.SPEAK": 3
+}
+```
+
+
+<b>appHMIType Priority Table</b><br>
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:117]:
```suggestion
<b>`appHMIType` Priority Table</b><br>
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+AppHMIType priority table describes the priority for each app type (`appHMIType`). When a conflict between RPCs with the same priority in the RPC priority table occurs, the RPC with the higher priority is determined according to AppHMIType priority table. Below is the default settings of AppHMIType priority table.
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:118]:
```suggestion
`AppHMIType` priority table describes the priority for each app type (`appHMIType`). When a conflict between RPCs with the same priority in the RPC priority table occurs, the RPC with the higher priority is determined according to `AppHMIType` priority table. Below are the default settings of `AppHMIType` priority table.
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+
+<b>Table 3.</b> Default settings of AppHMIType priority table
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:120]:
```suggestion
<b>Table 3.</b> Default settings of `AppHMIType` priority table
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+- AppHMIType priority table
+
+<b>RPC Priority Table</b><br>
+The RPC priority table describes the priority of each RPC. When multiple RPCs occur at the same time, the RPC with the higher priority is determined according to the RPC priority table. The target RPCs in this proposal are ONS and TTS. ONS RPCs include such as `Alert` and `PerformInteraction`. The RPC priority table can set the priority for each ONS RPC. On the other hand, there are two types of TTS RPCs, TTS with ONS and TTS only. TTS with ONS RPCs shall follow the processed priority of the ONS RPC, while TTS only RPCs can set the priority individually.
+
+Below is the default settings of RPC priority table.
+
+<b>Table 1.</b> Default settings of priority table
+
+| RPC_Name(String)        | Priority(INT) | Note(String)     |
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:42]:
```suggestion
| RPC_Name (String)        | Priority (INT) | Note (String)     |
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+| UI.Alert                | 2             |                  |
+| UI.PerformAudioPassThru | 2             |                  |
+| UI.PerformInteraction   | 3             |                  |
+| UI.ScrollableMessage    | 3             |                  |
+| UI.Slider               | 3             |                  |
+| TTS.Speak w/o ONS       | 3             | Lowest priority  |
+
+OEMs can modify the RPC priority table and adjust the RPC priority to their specifications.
+
+The table below shows how the RPC will be determined by the priority set during RPC conflict based on the default settings above (table 1).
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:54]:
```suggestion
The table below shows how the RPC will be determined by the priority set during RPC conflict based on the default settings above (Table 1).
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+}
+```
+
+
+<b>appHMIType Priority Table</b><br>
+AppHMIType priority table describes the priority for each app type (`appHMIType`). When a conflict between RPCs with the same priority in the RPC priority table occurs, the RPC with the higher priority is determined according to AppHMIType priority table. Below is the default settings of AppHMIType priority table.
+
+<b>Table 3.</b> Default settings of AppHMIType priority table
+
+| appHMIType(String) | Priority(INT) | Note(String)     |
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:122]:
```suggestion
| `appHMIType` (String) | Priority (INT) | Note (String)     |
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+|:-:                 |:-:            |:-:               |
+| Navigation         | 1             | Highest priority |
+| Projection         | 1             |                  |
+| Media              | 2             |                  |
+| Other              | 3             | Lowest priority  |
+
+Note that the "Other" in table 3 refers to appHMITypes other than Navigation, Projection and Media.
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:129]:
```suggestion
Note that the "Other" in Table 3 refers to `appHMITypes` other than Navigation, Projection and Media.
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+OEMs can modify the AppHMIType priority table and adjust the priority of application types according to their own specifications.
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:130]:
```suggestion
OEMs can modify the `AppHMIType` priority table and adjust the priority of application types according to their own specifications.
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+
+The table below shows how the RPC will be determined by the priority set during RPC conflict, based on the default settings above (table 3).
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:132]:
```suggestion
The table below shows how the RPC will be determined by the priority set during RPC conflict, based on the default settings above (Table 3).
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+
+<b>Table 4.</b> Priority result of Table 3
+
+![Table4_Priority_result_of_Table3.png](../assets/proposals/NNNN-Add-RPC-Conflict-Management/Table4_Priority_result_of_Table3.png)
+
+(* 2): According to the current SDL regulations, when a certain app has an HMI level of FULL, the HMI level of other similar apps will not be FULL. Therefore, there is a low possibility of RPC conflict with similar apps. Thus, RPC conflict for similar apps is outside the scope of this proposal.<br>
+(* 3): For the app types with the same priority, the first RPC will be prioritized.
+
+Below shows the Json example for the AppHMIType priority table:
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:141]:
```suggestion
Below shows the Json example for the `AppHMIType` priority table:
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+    "Navigation:": 1,
+    "Projection:": 1,
+    "Media": 2,
+    "Communication": 2,
+    "Other": 3
+}
+```
+
+#### RPC Conflict Management Module
+The following functions of InterruptManager are explained.
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:186]:
```suggestion
The following explains functions of `InterruptManager`.
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+
+<b>1.Loading of configuration file</b><br>
+InterruptManagerConfig is loaded during the startup of SDL Core.
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:189]:
```suggestion
`InterruptManagerConfig` is loaded during the startup of SDL Core.
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+![Figure3_TTS_RPCs_conflict.png](../assets/proposals/NNNN-Add-RPC-Conflict-Management/Figure3_TTS_RPCs_conflict.png)
+
+
+## Potential downsides
+None.
+
+
+## Impact on existing code
+
+1. Add InterruptManager to the SDL Core source code to manage RPC conflicts.
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:212]:
```suggestion
1. Add `InterruptManager` to the SDL Core source code to manage RPC conflicts.
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+2. Add InterruptManagerConfig that sets RPC priority to the SDL Core configuration file.
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:213]:
```suggestion
2. Add `InterruptManagerConfig` that sets RPC priority to the SDL Core configuration file.
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+3. Add guidelines for OEMs such as InterruptManagerConfig placement and setting method on SDLC official website.
```
Comment (theresalech) [inline proposals/NNNN-Add-RPC-Conflict-Management.md:214]:
```suggestion
3. Add guidelines for OEMs such as `InterruptManagerConfig` placement and setting method on SDL Developer Portal (smartdevicelink.com).
```

File: proposals/NNNN-Add-RPC-Conflict-Management.md (hunk 2 above)
Comment (Akihiro-Miyazaki) [reply proposals/NNNN-Add-RPC-Conflict-Management.md]: Hi. @theresalech -san,
Thank you for your review.
I will remove the word "SDL Documents" and make a commit it.
Could you please confirm it again?
Thank you.

**Comment by Akihiro-Miyazaki** [comment]
Hi. @theresalech -san,
Thank you in advance for your review.
I made a commit the PR file.
Could you please confirm them.
Thank you.

**Pull request merged**